### PR TITLE
feat(ethexe): Initial fast sync protocol

### PR DESCRIPTION
**Pull request by ark0f**

### CODE REVIEW

**Review by grishasobol — APPROVED** [review]
I don't find any problems with this PR, just some remarks. Please add TODO if you do not wanna to fix in this PR some of them.

File: ethexe/network/src/db_sync/ongoing.rs [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        let ongoing_request = self
+            .active_requests
+            .remove(&request_id)
+            .expect("unknown response");
+        let request_id = ongoing_request.request_id;
+
+        let new_ongoing_request = match ongoing_request.try_complete(peer, response) {
+            Ok(response) => return Ok((request_id, response)),
+            Err(new_ongoing_request) => new_ongoing_request,
+        };
```
[the remark this answers is not in the document]
Comment (ark0f) [reply ethexe/network/src/db_sync/ongoing.rs:278]: It is controlled by amount of rounds (unique requests to peers) or by timeout. Check `db_sync::Config`

File: ethexe/network/src/db_sync/ongoing.rs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+        &mut self,
+        behaviour: &mut InnerBehaviour,
+        peer: PeerId,
+        request_id: OutboundRequestId,
+        response: Response,
+    ) -> Result<(RequestId, Response), PeerResponse> {
+        let ongoing_request = self
+            .active_requests
+            .remove(&request_id)
+            .expect("unknown response");
```
[the remark this answers is not in the document]
Comment (ark0f) [reply ethexe/network/src/db_sync/ongoing.rs:272]: Each `request_id` is from `request_response::Behaviour` event, so it's impossible to have unknown response

File: ethexe/common/src/db.rs [hunk 3]
```diff
@@ -81,6 +81,7 @@ pub trait CodesStorage: Send + Sync {
 
     fn program_code_id(&self, program_id: ProgramId) -> Option<CodeId>;
     fn set_program_code_id(&self, program_id: ProgramId, code_id: CodeId);
+    fn program_ids(&self) -> BTreeSet<ProgramId>;
```
Comment (breathx) [inline ethexe/common/src/db.rs:84]: It seems that you could just iter through program_id to code_id, verifying both K and V on eth ?

File: ethexe/network/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+use std::{io, marker::PhantomData};
+
+pub(crate) struct ParityScaleCodec<Req, Resp>(PhantomData<(Req, Resp)>);
+
+impl<Req, Resp> ParityScaleCodec<Req, Resp> {
+    const MAX_REQUEST_SIZE: u64 = 1024 * 1024;
+    const MAX_RESPONSE_SIZE: u64 = 10 * 1024 * 1024;
+}
+
+#[async_trait]
```
Comment (breathx) [inline ethexe/network/src/utils.rs:34]: can't we avoid it?

File: ethexe/network/src/db_sync/mod.rs [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+};
+use parity_scale_codec::{Decode, Encode};
+use std::{
+    collections::{BTreeMap, BTreeSet, VecDeque},
+    task::{Context, Poll},
+    time::Duration,
+};
+
+const STREAM_PROTOCOL: StreamProtocol =
+    StreamProtocol::new(concat!("/ethexe/db-sync/", env!("CARGO_PKG_VERSION")));
```
Comment (breathx) [inline ethexe/network/src/db_sync/mod.rs:50]: only two identical nodes could be synced?

File: ethexe/network/src/db_sync/mod.rs [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+            [
+                (hash1, b"1".to_vec()),
+                (hash2, b"2".to_vec()),
+                (hash3, b"3".to_vec()),
+            ]
+            .into(),
+        );
+        assert_eq!(
+            request.validate_response(&response),
+            Err(RequestValidationError::ExcessiveData)
```
Comment (breathx) [inline ethexe/network/src/db_sync/mod.rs:573]: we can just strip it

File: ethexe/network/src/db_sync/mod.rs [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+    /// Request kind unequal to response kind
+    TypeMismatch,
+    /// Response contains more data than requested
+    ExcessiveData,
+    /// Hashed data unequal to its corresponding hash
+    DataHashMismatch,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub enum Request {
```
Comment (breathx) [inline ethexe/network/src/db_sync/mod.rs:69]: shouldn't we add block hash for which we're querying data?

**Review by breathx — APPROVED** [review]

File: ethexe/network/src/db_sync/mod.rs (hunk 5 above)
Comment (ukint-vs) [reply ethexe/network/src/db_sync/mod.rs:50]: better to use constant version and bump when needed.

**Review by ukint-vs — APPROVED** [review]

File: ethexe/network/src/utils.rs (hunk 4 above)
Comment (ark0f) [reply ethexe/network/src/utils.rs:34]: We can't. It's `libp2p` requirement

File: ethexe/network/src/db_sync/mod.rs (hunk 5 above)
Comment (ark0f) [reply ethexe/network/src/db_sync/mod.rs:50]: Anyway, version management is for the future

File: ethexe/network/src/db_sync/mod.rs (hunk 7 above)
Comment (ark0f) [reply ethexe/network/src/db_sync/mod.rs:69]: Data we request is block hash agnostic

File: ethexe/common/src/db.rs (hunk 3 above)
Comment (breathx) [reply ethexe/common/src/db.rs:84]: tbd later

**Pull request merged**